cheatsheets/Password_Storage_Cheat_Sheet.md

```diff
@@ -1,567 +1,151 @@
+- **Use a [pepper](#peppering).**
 
-## Do not limit the character set and set long max lengths for credentials
+## Hashing vs Encryption
 
-Some organizations restrict the 1) types of special characters and 2) length of credentials accepted by systems because of their inability to prevent SQL Injection, Cross-site scripting, command-injection and other forms of injection attacks. These restrictions, while well-intentioned, facilitate certain simple attacks such as brute force.
+Hashing and encrypted are two terms that are often confused or used incorrectly. The key difference between them is that hashing is a **one way** function (i.e, it is not possible to "decrypt" a hash and obtain the original value), whereas encryption is a two-way function.
 
-Do not allow short or no-length passwords and do not apply character set, or encoding restrictions on the entry or storage of credentials. Continue applying encoding, escaping, masking, outright omission, and other best practices to eliminate injection risks.
+In almost all circumstances passwords should be hashed, rather than encrypted, as this makes it difficult or impossible for an attacker to obtain the original passwords from the hashes.
 
-A reasonable long password length is 160. Very long password policies can [lead to DoS in certain circumstances](http://arstechnica.com/security/2013/09/long-passwords-are-good-but-too-much-length-can-be-bad-for-security/).
+Encryption should only be used in edge cases where it is necessary to be able to obtain the original password. Some examples of where this might be necessary are:
 
-## Hash the password as one of several steps
+- If the application needs to use the password to authenticate against an external legacy system that doesn't support SSO.
+- If it is necessary to retrieve individual characters from the password.
 
-Very large passwords can be a performance bottleneck or a DoS situation when users utilize very long passwords. Also, some implementations of some of the adaptive algorithms suggested below, such as bcrypt, truncate long passwords making them less effective. By first hashing the passwords with a fast hash such as SHA-512, and then hashing the output with a slower and more secure hash such as bcrypt, even giant passwords are reduced to 512 bits, solving both problems.
+The ability to decrypt passwords represents a serious security risk, so should be fully risk assessed. Where possible an alternative architecture should be used to avoid the need to store passwords in an encrypted form.
 
-## Use a cryptographically strong credential-specific salt
+This Cheat Sheet is focused on password hashing - for further guidance on encrypting passwords see the [Cryptographic Storage Cheat Sheet](Cryptographic_Storage_Cheat_Sheet.md).
 
-A salt is fixed-length cryptographically-strong random value. Append credential data to the salt and use this as input to a protective function. 
+## How Attackers Crack Password Hashes
```

Comment:
It's really just there to give some background and understanding - this felt like the most natural place to put it, so the flow of the CS is:

- Background and context
- Key concepts
- Hashing algorithms

Happy to move this around though. @mackowski do you have any thoughts on this?

Reply:
Since we are focusing on builders, I generally look at it as such:
- Key focus and what the developer needs
- Best practices
- More thoughts on the topic at hand. They came here for the CS, if they want more, they'll look at it. In this case background, context, some general threats, etc.

Reply:
I think that it can stay here. For me this is a good format for this CS. This is logical point after the `Hashing vs Encryption` part. Especially because there is ToC so if one is not interested they can jump to the next section.